File: SDE_sheet/Arrays/24_longest_substring_without_repeat/solution.js
var lengthOfLongestSubstring = function(s) {
    let left = 0;
    let max = 0;
    const set = new Set();

    for (let right = 0; right < s.length; right++) {
        const char = s[right];

        while (set.has(char)) {
            set.delete(s[left]);
            left++;
        }

        set.add(char);
        max = Math.max(max, right - left + 1);
    }

    return max;
};

File: SDE_sheet/Arrays/24_longest_substring_without_repeat/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0;
        int max = 0;
        HashSet<Character> set = new HashSet<>();

        for (int right = 0; right < s.length(); right++) {
            char ch = s.charAt(right);

            while (set.contains(ch)) {
                set.remove(s.charAt(left));
                left++;
            }

            set.add(ch);
            max = Math.max(max, right - left + 1);
        }

        return max;
    }
}
